libvast/include/vast/cast.hpp

```diff
@@ -0,0 +1,266 @@
+                                             can_cast_field_types.error()));
+        }
+      }
+    }
+    return {};
+  }
+
+  static auto
+  cast(const record_type& from_type,
+       std::shared_ptr<type_to_arrow_array_t<record_type>> from_array,
+       const record_type& to_type) noexcept
+    -> std::shared_ptr<type_to_arrow_array_t<record_type>> {
+    VAST_ASSERT(can_cast(from_type, to_type));
+    if (from_type == to_type)
+      return from_array;
+    // NOLINTNEXTLINE
+    auto impl = [&](const auto& impl, const record_type& to_type,
+                    std::string_view key_prefix) noexcept
+      -> std::shared_ptr<type_to_arrow_array_t<record_type>> {
+      auto fields = arrow::FieldVector{};
+      auto children = arrow::ArrayVector{};
+      fields.reserve(to_type.num_fields());
+      children.reserve(to_type.num_fields());
+      for (const auto& to_field : to_type.fields()) {
+        const auto key = key_prefix.empty()
+                           ? std::string{to_field.name}
+                           : fmt::format("{}.{}", key_prefix, to_field.name);
+        fields.push_back(to_field.type.to_arrow_field(to_field.name));
+        if (const auto* r = caf::get_if<record_type>(&to_field.type)) {
+          children.push_back(impl(impl, *r, key));
+          continue;
+        }
+        const auto index = from_type.resolve_key(key);
+        if (!index) {
+          // The field does not exist, so we insert a bunch of nulls.
+          children.push_back(
+            arrow::MakeArrayOfNull(to_field.type.to_arrow_type(),
+                                   from_array->length())
+              .ValueOrDie());
+          continue;
+        }
+        // The field exists, so we can insert the casted column.
+        children.push_back(cast_helper<type, type>::cast(
+          from_type.field(*index).type,
+          static_cast<arrow::FieldPath>(*index).Get(*from_array).ValueOrDie(),
+          to_field.type));
+      }
+      return type_to_arrow_array_t<record_type>::Make(children, fields)
+        .ValueOrDie();
+    };
+    return impl(impl, to_type, "");
+  }
+};
+
+} // namespace detail
+
+/// Determines whether a cast from *from_type* to *to_type* is possible.
+/// @returns Nothing on success, or an error detailing why the cast is not
+/// possible on failure.
+template <type_or_concrete_type FromType, type_or_concrete_type ToType>
+auto can_cast(const FromType& from_type, const ToType& to_type) noexcept
+  -> caf::expected<void> {
+  return detail::cast_helper<FromType, ToType>::can_cast(from_type, to_type);
+}
+
+/// Casts a table slice to another schema.
+/// @param from_slice The table slice to cast.
+/// @param to_schema The schema to cast towards.
+/// @pre can_cast(from_slice.schema(), to_schema)
+/// @post result.schema() == to_schema
+/// @returns A slice that exactly matches *to_schema*.
+auto cast(table_slice from_slice, const type& to_schema) noexcept
+  -> table_slice;
```

Comment:
```suggestion
/// Casts a table slice to another schema.
/// @param from The table slice to cast.
/// @param to The schema to cast towards.
/// @pre can_cast(from_slice.schema(), to_schema)
/// @post result.schema() == to_schema
/// @returns A slice that exactly matches *to_schema*.
auto cast(table_slice from, const type& to) noexcept -> table_slice;
```

To me, "cast" already implies what's going on. It doesn't need to be duplicated into the prefix of the argument identifiers.

(Bikeshedding comment, because everything else looks great.)

Reply:
I'm against this change. The implementation of this function is much easier to read with the very explicit variable names, and the variable names in the function definition should match those in its definition.